### ISSUE TITLE
Policy on enrollment no longer working in `2024.4` resulting in `UserSourceConnection has no user.`

### ISSUE THREAD

**Issue by ToshY** [goauthentik/authentik#9477]
**Describe the bug**

With a expresion policy in place to allow only certain Google users to login, it now shows the message `UserSourceConnection has no user.` (disregarding if it passes/fails the policy) due to failing to run a policy which worked fine in `2024.2.3`.

**To Reproduce**

1. Add Google as social login according to docs: https://docs.goauthentik.io/integrations/sources/google/
2. Create an expression policy with the following expression:

`google-email-expression`
```python
if context["source"].provider_type != "google":
    return True

email = request.context["prompt_data"]["email"]
# Direct set username to email
request.context["prompt_data"]["username"] = email
# Set username to email without domain
request.context["prompt_data"]["username"] = email.split("@")[0]

accepted_users = ["user1@gmail.com","user2@gmail.com"]

# Match username
user_matched = any(user == email for user in accepted_users)
if not user_matched:
    ak_message("You are not allowed to log in. This attempt has been logged.")
return user_matched
```

3. Add the above policy to the `default-source-enrollment`.

![image](https://github.com/goauthentik/authentik/assets/31921460/0e26b82c-4fbf-4c2b-aca1-b5a0eb73a980)

4. Try to login with Google as `user1@gmail.com` (allowed according to policy)

![image](https://github.com/goauthentik/authentik/assets/31921460/2a986a2d-a162-4d61-a4a4-d686d761d5ac)

5. Try to login with Google as `user3@gmail.com` (not allowed according to policy)

![image](https://github.com/goauthentik/authentik/assets/31921460/72b4dca7-e0c7-40f0-9df7-64bd59547602)

**Expected behavior**

Same behaviour as `2024.2.3`, account is either created for allowed users passing the policy, or the message is displayed that the user is not allowed.

![image](https://github.com/goauthentik/authentik/assets/31921460/0e14e3de-b92f-4651-9abe-da2f7ddeaf87)

**Screenshots**

See above.

**Logs**

```json
{
  "auth_via": "unauthenticated",
  "domain_url": "auth.example.com",
  "event": "Policy failed to run",
  "exc": "Traceback (most recent call last):\n  File \"/authentik/policies/process.py\", line 144, in run\n    self.connection.send(self.profiling_wrapper())\n                         ^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/authentik/policies/process.py\", line 139, in profiling_wrapper\n    return self.execute()\n           ^^^^^^^^^^^^^^\n  File \"/authentik/policies/process.py\", line 90, in execute\n    self.create_event(\n  File \"/authentik/policies/process.py\", line 60, in create_event\n    event = Event.new(\n            ^^^^^^^^^^\n  File \"/authentik/events/models.py\", line 218, in new\n    cleaned_kwargs = cleanse_dict(sanitize_dict(kwargs))\n^^^^^^^^^^^^^^^^^^^^^\nFile\"/authentik/events/utils.py\", line 191, in sanitize_dict\n    new_value = sanitize_item(value)\n                ^^^^^^^^^^^^^^^^^^^^\n  File \"/authentik/events/utils.py\", line 114, in sanitize_item\n    return sanitize_dict(value)\n           ^^^^^^^^^^^^^^^^^^^^\n  File \"/authentik/events/utils.py\", line 191, in sanitize_dict\n    new_value = sanitize_item(value)\n                ^^^^^^^^^^^^^^^^^^^^\n  File \"/authentik/events/utils.py\", line 114, in sanitize_item\n    return sanitize_dict(value)\n           ^^^^^^^^^^^^^^^^^^^^\n  File \"/authentik/events/utils.py\", line 191, in sanitize_dict\n    new_value = sanitize_item(value)\n                ^^^^^^^^^^^^^^^^^^^^\n  File \"/authentik/events/utils.py\", line 127, in sanitize_item\n    return sanitize_dict(model_to_dict(value))\n                         ^^^^^^^^^^^^^^^^^^^^\n  File \"/authentik/events/utils.py\", line 66, in model_to_dict\n    name = str(model)\n           ^^^^^^^^^^\n  File \"/authentik/core/models.py\", line 635, in __str__\n    return f\"User-source connection (user={self.user.username}, source={self.source.slug})\"\n                                           ^^^^^^^^^\n  File \"/ak-root/venv/lib/python3.12/site-packages/django/db/models/fields/related_descriptors.py\", line 264, in __get__\n    raise self.RelatedObjectDoesNotExist(\nauthentik.core.models.RelatedObjectDoesNotExist: UserSourceConnection has no user.",
  "host": "auth.example.com",
  "level": "warning",
  "logger": "authentik.policies.process",
  "pid": 88,
  "request_id": "e20f015d8b0a4447887685e1ebd7720e",
  "schema_name": "public",
  "timestamp": "2024-04-26T21:34:05.375936"
}
```

**Version and Deployment (please complete the following information):**

-   authentik version: `2024.4.0`, `2024.4.1`.
-   Deployment: `docker compose v2.26.1`

**Comment by BeryJu** [comment]
The actual bug I can confirm, however this issue is only triggered by having execution logging enabled in the policy. Once that is disabled this error shouldn't happen anymore

**Comment by ToshY** [comment]
Hey @BeryJu 👋 

Thanks for the quick response. I totally forgot to mention that detail in my policy, but you're right, I have the logging enabled. I just disabled it and works like you said, so I'm totally fine with that workaround for the time being.

**Comment by ToshY** [comment]
Hey @BeryJu 👋 

Maybe you've already taken this into account in #9480, but I just tested that if I add another expression policy (e.g. `custom-enrollment-group`) with execution logging disabled, and bind the policy to the `default-source-enrollment` as well, I get both the policy message that user is not allowed to login (from the previous `google-email-expression` policy), but also again the error `UserSourceConnection has no user.` probably originating from the new policy.

`google-email-expression` (existing policy with execution logging disabled)

![image](https://github.com/goauthentik/authentik/assets/31921460/5cfd22ce-032a-4fd2-8fd2-a45de392c283)

`custom-enrollment-group` (new policy with execution logging disabled)

![image](https://github.com/goauthentik/authentik/assets/31921460/0fb808ea-ef6c-4401-aa26-2e0880fea9c6)

> Note: group `app-user` exists

`default-source-enrollment`

![image](https://github.com/goauthentik/authentik/assets/31921460/7c8b9332-1c16-427e-9f94-9bb9f9b48ad7)


Response for `user1@gmail.com` (allowed)

![image](https://github.com/goauthentik/authentik/assets/31921460/1ddad44e-20d3-498f-8019-5dcf8119dce7)

Response for `user3@gmail.com` (not allowed)

![image](https://github.com/goauthentik/authentik/assets/31921460/a22e16fc-e8e9-4319-9d3a-cb647b8a5de7)

---

TLDR; I still get the `UserSourceConnection has no user.` if the execution logging is disabled on new policies added to the enrollment.

- Tried removing the `custom-enrollment-group` and recreate it; didn't help
- Tried to down/up my stack; didn't help
- Tried `Clear cache` button on the policies; didn't help

If you have any ideas on why this still happens on newly added policies, please let me know.
